system/lib/wasmfs/file.h

```diff
@@ -0,0 +1,32 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the file object of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#pragma once
+
+#include <emscripten/html5.h>
+#include <mutex>
+#include <vector>
+#include <wasi/api.h>
+
+namespace wasmfs {
+
+class File {
+  // TODO: Add other File properties later.
+
+  // A mutex is needed for multiple accesses to the same file.
+  std::mutex mutex;
+
+public:
+  virtual __wasi_errno_t read(
+    const __wasi_iovec_t* iovs, size_t iovs_len, __wasi_size_t* nread) = 0;
+  virtual __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) = 0;
```

Comment:
What do we think about going to the model I mentioned yesterday, of File being the base, on top of which we have Data, Directory, and Link? If so, then the read and write methods could only be on Data. That seems cleaner, as otherwise we'd want to put all the directory and link methods on File too..?

That doesn't need to block this PR though. But when we add Directory or Data we should make a decision and refactor.

Reply:
That does make sense. I agree that we should add that in the next PR. I think it will make a lot of sense when we have Directory as well.